File: Data Structures/Java/Disjoint_Set_Union.java
// By : Mohamed Bakr

public class Disjoint_Set_Union 
{
	public int[] parent;
	public int[] sz;
	public Disjoint_Set_Union(int n)
	{
		parent = new int[n+1];
		sz = new int[n+1];
		for(int i = 0;i<=n;i++)
		{
			parent[i] = i;
			sz[i] = 1;
		}
	}
	public int find(int u)
	{
		if (u == parent[u]) return u;
		return parent[u] = find(parent[u]);
	}
	public boolean add(int u, int v)
	{
		u = find(u); v = find(v);
		if (u == v) return false;
		if (sz[u] < sz[v]) 
		{
			parent[u] = v;
			sz[v] += sz[u];
		}
		else 
		{
			parent[v] = u;
			sz[u] += sz[v];
		}
		return true;
	}
	public boolean same(int u, int v)
	{ 
		return find(u) == find(v); 
	}
}

File: Data Structures/Python/Disjoint_Set_Union.py
# By : Mohamed Bakr

class Disjoint_Set_Union:
    def __init__(self, n):
        self.parent = [0] * (n+1)
        self.sz = [1] * (n+1)
        for i in range(0,n+1):
            self.parent[i] = i
            
    def find(self, u):
        if u == self.parent[u]:
            return u
        else:
            self.parent[u] = self.find(self.parent[u])
            return self.parent[u]

    def add(self, u, v):
        u = self.find(u)
        v = self.find(v)
        if u == v:
            return False
        if self.sz[u] < self.sz[v]:
            self.parent[u] = v
            self.sz[v] += self.sz[u]
        else:
            self.parent[v] = u
            self.sz[u] += self.sz[v]
        return True

    def same(self, u, v):
        return self.find(u) == self.find(v)
